### PR TITLE
feat: introduce proxy related variables

### PR DESCRIPTION
**Pull request by shipperizer**
# Description

Introducing support for HTTP proxying

Golang respects the usage of the environment variables (below) when using HTTP clients:

The following variables are set in the  environment context/shell to constraint the `lego` command
* `HTTP_PROXY`
* `HTTPS_PROXY`
* `NO_PROXY`

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I Have bumped the version of the library

### CODE REVIEW

File: lib/charms/lego_base_k8s/v0/lego_client.py
Comment (gruyaume) [inline lib/charms/lego_base_k8s/v0/lego_client.py:1]: We should bump the lib version

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/lego_base_k8s/v0/lego_client.py [hunk 1]
```diff
@@ -316,6 +319,25 @@ def _cmd(self) -> List[str]:
...
+        """
+        common_config = {}
+        if self._http_proxy:
+            common_config["HTTP_PROXY"] = self._http_proxy
+        if self._https_proxy:
+            common_config["HTTPS_PROXY"] = self._https_proxy
+        if self._no_proxy:
+            common_config["NO_PROXY"] = self._no_proxy
+
+        return common_config
```
Comment (saltiyazan) [inline lib/charms/lego_base_k8s/v0/lego_client.py:339]:
```suggestion
 def _get_app_environment(self) -> Dict[str, str]:
        """Extract proxy model environment variables."""
        env = {}
        http_proxy = os.environ.get("HTTP_PROXY", "")
        https_proxy = os.environ.get("HTTPS_PROXY", "")
        no_proxy = os.environ.get("NO_PROXY", "")
        if http_proxy and https_proxy:
            logger.info(
                "Proxy settings found: HTTP_PROXY=%s, HTTPS_PROXY=%s, NO_PROXY=%s",
                http_proxy, https_proxy, no_proxy
            )
            env["HTTP_PROXY"] = http_proxy
            env["HTTPS_PROXY"] = https_proxy
            env["NO_PROXY"] = no_proxy

        return env
```

Comment (saltiyazan) [inline lib/charms/lego_base_k8s/v0/lego_client.py:339]:
```suggestion
def _get_app_environment(self) -> Dict[str, str]:
        """Extract proxy model environment variables."""
        env = {}
        if (http_proxy := os.environ.get("HTTP_PROXY", "")):
            env["HTTP_PROXY"] = http_proxy
        if (https_proxy := os.environ.get("HTTPS_PROXY", "")):
            env["HTTPS_PROXY"] = https_proxy
        if(no_proxy := os.environ.get("NO_PROXY", "")):
            env["NO_PROXY"] = no_proxy
        return env
```

File: lib/charms/lego_base_k8s/v0/lego_client.py [hunk 2]
```diff
@@ -358,3 +380,24 @@ def _server(self) -> Optional[str]:
...
+        if not isinstance(https_proxy, str):
+            return None
+        return https_proxy
+
+    @property
+    def _no_proxy(self) -> Optional[str]:
+        no_proxy = self.model.config.get("no_proxy", None)
+        if not isinstance(no_proxy, str):
+            return None
+        return no_proxy
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply lib/charms/lego_base_k8s/v0/lego_client.py]: Changed

File: lib/charms/lego_base_k8s/v0/lego_client.py
Comment (saltiyazan) [reply lib/charms/lego_base_k8s/v0/lego_client.py:1]: Done

File: lib/charms/lego_base_k8s/v0/lego_client.py [hunk 3]
```diff
@@ -316,6 +320,22 @@ def _cmd(self) -> List[str]:
...
+    @property
+    def _app_environment(self) -> Dict[str, str]:
+        """Extract proxy model environment variables."""
+        env = {}
+
+        if (http_proxy := get_env_var(env_var="HTTP_PROXY")):
+            env["HTTP_PROXY"] = http_proxy
+        if (https_proxy := get_env_var(env_var="HTTPS_PROXY")):
+            env["HTTPS_PROXY"] = https_proxy
+        # there's no need for no_proxy if there's no http_proxy or https_proxy
```
Comment (saltiyazan) [inline lib/charms/lego_base_k8s/v0/lego_client.py:334]: I added the extra condition `and (http_proxy and https_proxy):` following the example in github-runner-operator [here](https://github.com/canonical/github-runner-operator/blob/82067e501bbb9e7e23330c433098b4083beaa433/src/charm_state.py#L688).

File: lib/charms/lego_base_k8s/v0/lego_client.py [hunk 4]
```diff
@@ -316,6 +320,22 @@ def _cmd(self) -> List[str]:
...
         ]
 
+
+
+    @property
+    def _app_environment(self) -> Dict[str, str]:
+        """Extract proxy model environment variables."""
+        env = {}
+
+        if (http_proxy := get_env_var(env_var="HTTP_PROXY")):
```
Comment (gruyaume) [inline lib/charms/lego_base_k8s/v0/lego_client.py:330]: Is this the right env var that we would see? Looking at the github runner app, the env variable is `JUJU_CHARM_HTTP_PROXY`, not `HTTP_PROXY`

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/lego_base_k8s/v0/lego_client.py (hunk 4 above)
Comment (saltiyazan) [reply lib/charms/lego_base_k8s/v0/lego_client.py]: You are right, fixed

**Review by gruyaume — APPROVED** [review]
Looks good, @shipperizer please test the change before merging this here to avoid back and forth

**Comment by saltiyazan** [comment]
Tested and worked locally. Merging.

**Pull request merged**